umpscqueue.go

```diff
@@ -0,0 +1,162 @@
+}
+
+// This value is chose arbitrarily, as increasing it gives diminishing returns. With some testing (on
+// 64-core machines), when the segment size is smaller than 2^10, the queue becomes slower as
+// parallelism increases, while there is no statistically significant difference beyond 2^12.
+const segmentSize = 1 << 12
+
+// key/value pair representing the [time.Duration] that should be inserted into the backing
+// [metrics.TimeHistogram], and whether the value is ready to be read. The reading goroutine should
+// not attempt to read the value until the ready [sync.WaitGroup] has been marked as done.
+type queueValue[T any] struct {
+	value T
+	ready sync.WaitGroup
+}
+
+// init initializes the [sync.WaitGroup] so that get blocks until set is called.
+func (hv *queueValue[T]) init() {
+	hv.ready.Add(1)
+}
+
+// set sets the value and marks it as ready.
+func (hv *queueValue[T]) set(value T) {
+	hv.value = value
+	hv.ready.Done()
+}
+
+// get waits for the value to be ready, then reads it.
+func (hv *queueValue[T]) get() T {
+	hv.ready.Wait()
+	return hv.value
+}
+
+type queueSegment[T any] struct {
+	// Incremented every time a writer wants to write to this segment, and prevents multiple writers from
+	// attempting to write to the same index. If the index is greater than the size of the segment,
+	// pending writers should try again in the next segment.
+	idx atomic.Uint64
+	// Padding to prevent false sharing.
+	_ [cacheLineSize - unsafe.Sizeof(atomic.Uint64{})]byte
+	// The set of values this segment.
+	values []queueValue[T]
+	// Synchronizes the creation of the next segment.
+	nextOnce sync.Once
+	next     *queueSegment[T]
+}
+
+// newSegment creates a new queueSegment and pre-allocates the value slice by either reusing one
+// from the pool or creating a fresh one.
+func (q *UMPSCQueue[T]) newSegment() *queueSegment[T] {
+	values, ok := q.segmentPool.Get().([]queueValue[T])
+	if !ok {
+		values = make([]queueValue[T], segmentSize)
+	}
+	for i := range values {
+		values[i].init()
+	}
+
+	s := &queueSegment[T]{
+		values: values,
+	}
+	// Storing math.MaxUint64 means the first call to Add(1) will return 0, not 1! Otherwise, it becomes
+	// very messy as the real intended index is actually s.idx.Add(1) - 1. Setting this once makes this
+	// less error-prone.
+	s.idx.Store(math.MaxUint64)
+	return s
+}
+
+func (q *UMPSCQueue[T]) loadNext(s *queueSegment[T]) *queueSegment[T] {
+	s.nextOnce.Do(func() {
+		s.next = q.newSegment()
+	})
+	return s.next
+}
+
+// Take returns the next value in the queue, blocking if it is empty. It is not safe to invoke Take
+// from multiple goroutines.
+//
+// As it completes reading a segment, it returns the backing value slice to the pool. The actual
+// queueSegment itself cannot be reused as it contains the pointer to the next segment, which cannot
+// safely be updated as it cannot be determined whether all writers have released all references to
+// it.
+func (q *UMPSCQueue[T]) Take() T {
```

Comment:
`Dequeue` and `Enqueue` would be more consistent names for the `Take` and `Add` methods since that's how these operations are called in remaining queues.